[2단계 - 리팩터링] 엠제이 (최민준) 미션 제출합니다.

theminjunchoi opened the pull request:
안녕하세요 강민햄 반가워요. 😝
저는 포스트말론의 overdrive를 참 좋아하는데요,
강민햄의 최애곡은 뭔가여?

step2 요구사항을 보니까 특별히 더 할 게 없어보이더라고요?
그래서 중복 코드 제거하고 지난 pr에서 남겨준 리뷰 반영해봤습니다!

추가로 가변인자를 PreparedStatementSetter로 변환하는 어댑터를 만들어봤어요~

편하게 리뷰해주세용~

<img width="259" height="194" alt="image" src="https://github.com/user-attachments/assets/c9f4fef9-061a-4c06-bf3b-b69462e5025c" />

theminjunchoi committed:
refactor: try-with-resources로 JDBC 리소스 처리 코드 개선

- `try-with-resources` 구문을 사용해 커넥션 및 리소스 자동 해제 적용
- 명시적인 리소스 해제 코드 제거
- 코드 가독성 및 안정성 향상

theminjunchoi committed:
feat: PreparedStatementSetter 및 관련 메서드 추가

- `PreparedStatementSetter` 인터페이스 추가: PreparedStatement 값 설정 지원
- `JdbcTemplate` 클래스에 `update`, `queryForObject`, `query` 메서드 오버로드 추가: PreparedStatementSetter 활용
- `UserDaoTest`에 PreparedStatementSetter를 이용한 테스트 케이스 추가:
  - `updateUsingPreparedStatementSetter`
  - `findByIdUsingPreparedStatementSetter`
- 코드 재사용성과 가독성 향상

theminjunchoi committed:
refactor: JdbcTemplate 메서드 구조 개선 및 공통 로직 분리

- `createPreparedStatementSetter` 메서드 추가: PreparedStatementSetter 객체 생성 간소화
- `extractResults` 메서드 추가: ResultSet 처리 로직 분리 및 재사용성 확보
- `getSingleResult` 메서드 추가: 단일 결과 반환 로직 재사용 가능하도록 분리
- `query`, `update`, `queryForObject` 메서드 간 코드 중복 제거
- 가독성 및 유지보수성 향상

ykmxxi commented on jdbc/src/main/java/com/interface21/jdbc/core/JdbcTemplate.java:
```diff
@@ -25,87 +25,82 @@ public JdbcTemplate(final DataSource dataSource) {
-            throw new EmptyResultDataAccessException("Query returned no results");
-        }
-        
-        if (results.size() > 1) {
-            throw new IncorrectResultSizeDataAccessException(1, results.size());
-        }
-        
-        return results.getFirst();
+        return queryForObject(sql, createPreparedStatementSetter(args), rowMapper);
+    }
+
+    public <T> T queryForObject(String sql, PreparedStatementSetter pss, RowMapper<T> rowMapper) {
+        List<T> results = query(sql, pss, rowMapper);
+        return getSingleResult(results);
```
각각의 메서드(`query`, `update`, `queryForObject`) 모두 두 가지 방식으로 구현해 제공한 이유가 있을까요?

theminjunchoi replied:
맞습니당

단순 가변인자로 바로 넘길 수 있게 만든 간편한 버전과
복잡한 파라미터 바인딩이 필요할 때를 위해 유연한 버전을 만들어놨는데,

사용자 편의성이랑 유연성 둘 다 제공하고 싶었어용

ykmxxi replied:
이해했어용 역시 천상계

ykmxxi commented on jdbc/src/main/java/com/interface21/jdbc/core/JdbcTemplate.java:
```diff
@@ -25,87 +25,82 @@ public JdbcTemplate(final DataSource dataSource) {
-            conn = dataSource.getConnection();
-            pstmt = conn.prepareStatement(sql);
-            
+        try (Connection conn = dataSource.getConnection();
+             PreparedStatement pstmt = conn.prepareStatement(sql)) {
```
저번 리뷰를 반영해주셨군요 👍 💯

theminjunchoi replied:
맞아여

ykmxxi reviewed (changes requested):
<img width="460" height="424" alt="image" src="https://github.com/user-attachments/assets/cf4763ff-5ddd-4980-80e6-1e22f5a88ee0" />

참 빠르시네요 민준씌 🚀  요새 가을냄새가 나기 시작해서 아침 저녁으로 쌀쌀한데 감기 조심하세용
참고로 제 최애곡은 전곡(다 명곡임)입니다 ㅎ 요즘 자주 듣는 곡은 Wrapped Around Your Finger

몇 가지 질문만 남겼어요. 역싀 천상계 답게 잘하시네요

그런데 이전 스텝에서 놓치고간 부분이 있어 description에도 질문 남겨요. 질문에 답변해주시면 다음 요청때 Approve & Merge 할께용 🙇 

```java
@FunctionalInterface
public interface RowMapper<T> {

    T mapRow(ResultSet rs, int rowNum) throws SQLException;
}
```

- `mapRow` 메서드의 `rowNum` 파라미터는 사용하지 않고 있는데 필요한 이유가 있을까요??

```java
public class UserDao {

    ...
    private RowMapper<User> userRowMapper() {
        return (rs, rowNum) -> new User(
                rs.getLong("id"),
                rs.getString("account"),
                rs.getString("password"),
                rs.getString("email")
        );
    }
}
```

- `userRowMapper()`를 통해 매번 새로운 매퍼 인스턴스를 생성할 필요가 있나요?? 상수로 두어 재사용할 수 있지 않을까요??

theminjunchoi committed:
refactor: RowMapper 객체 상수화

- `USER_ROW_MAPPER` 상수 추가: 중복 매핑 로직 제거
- `findAll`, `findById`, `findByAccount` 메서드에서 중복 매핑 코드 제거
- 유지보수성 및 가독성 향상

theminjunchoi committed:
refactor: RowMapper 인터페이스 단순화 및 관련 코드 수정

- `RowMapper` 메서드 시그니처에서 `rowNum` 매개변수 제거
- `JdbcTemplate`에서 `extractResults` 메서드 내부 로직 수정: `rowNum` 기반 처리 제거
- `UserDao` 및 관련 테스트(`UserDaoTest`)의 `RowMapper` 구현 코드 수정
- 불필요한 매개변수 제거로 코드 간소화 및 가독성 향상

theminjunchoi reviewed (commented):
> mapRow 메서드의 rowNum 파라미터는 사용하지 않고 있는데 필요한 이유가 있을까요??

첨엔 너무 익숙한 모양이라 아무 생각없이 만든 것 같네요 ㅎㅎ
저도 궁금해서 jdbc 라이브러리를 좀 찾아봤는데, 별 용도가 없는 것 같아요.
기껏해야 로깅정도?

클로드 선생님한테도 물어봤는데 딱히 용도가 없는 것 같아요
이게 왜 존재하는 걸까요?? 강민햄은 아시나요?

우선 지금은 안쓰고 있어서 제거해버렸습니다~!

> userRowMapper()를 통해 매번 새로운 매퍼 인스턴스를 생성할 필요가 있나요?? 상수로 두어 재사용할 수 있지 않을까요??

인정합니다. 반영했어용~
해피 추석보내세여~~
<img width="304" height="166" alt="image" src="https://github.com/user-attachments/assets/74f1fa1f-e652-49cc-8339-a5414d018717" />

ykmxxi commented:
저도 rowNum이 왜 존재하는지 저도 잘 모르겠네용. 찾아보니 아래 이슈에서 rowNum 파라미터의 필요성에 대해 과거에 논의가 이루어진 것 같아요.

#https://github.com/spring-projects/spring-framework/issues/7796

정리해보면 다음과 같네요
- 초기 설계 철학을 맞추기 위해: "더 많은 컨텍스트를 제공하는 게 낫다"는 원칙으로 처음부터 rowNum을 포함시킴
- 호환성 문제 (제거 불가능): 이미 공개된 인터페이스라서 제거하면 기존 코드가 모두 깨짐
